### PR TITLE
RMC-138 Scheduled Reminder Letters

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
Action scheduler needs to schedule remind letters.

# What has changed
* Add action types for reminder letters
* Create new UAC QID pair for reminder pack codes
* Calculate questionnaire type from treatment code
* Unit tests for calculating questionnaire type and requesting new UAC QID pairs

# How to test?
Run a maven build on this branch, build the linked branch of print file service then run the linked branch of acceptance tests

# Links
https://trello.com/c/9WfpfHYE/721-rmc-138-schedule-generate-and-distribute-reminder-letters-5
https://github.com/ONSdigital/census-rm-print-file-service/pull/11
https://github.com/ONSdigital/census-rm-acceptance-tests/pull/112

### CODE REVIEW

**Comment by AdamHawtin** [comment]
[RMC-138 Schedule, generate and distribute reminder letters (5)](https://trello.com/c/9WfpfHYE/721-rmc-138-schedule-generate-and-distribute-reminder-letters-5)

File: src/test/java/uk.gov.ons.census.action/messaging/FulfilmentRequestReceiverIT.java [hunk 1]
```diff
@@ -49,7 +49,7 @@
   private String outboundPrinterQueue;
 
   @Rule
-  public WireMockRule wireMockRule = new WireMockRule(wireMockConfig().port(8089).httpsPort(8443));
+  public WireMockRule mockCaseApi = new WireMockRule(wireMockConfig().port(8089).httpsPort(8443));
```
Comment (NickGrantONS) [inline src/test/java/uk.gov.ons.census.action/messaging/FulfilmentRequestReceiverIT.java]: might as well get rid of the https port seeing as it's unused

File: src/test/java/uk/gov/ons/census/action/builders/PrintFileDtoBuilderTest.java [hunk 2]
```diff
@@ -85,10 +87,10 @@ private PrintFileDto getExpectedPrintFileDto(Case caze) {
 
     printFileDto.setCaseRef(caze.getCaseRef());
 
-    // TODO: where are these stored and used?
-    printFileDto.setTitle("");
-    printFileDto.setForename("");
-    printFileDto.setSurname("");
+    //    // TODO: where are these stored and used?
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/census/action/builders/PrintFileDtoBuilderTest.java]: double commented but doesn't really matter

File: src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+import uk.gov.ons.census.action.model.repository.UacQidLinkRepository;
+
+@ContextConfiguration
+@SpringBootTest
+@ActiveProfiles("test")
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+@RunWith(SpringJUnit4ClassRunner.class)
+public class ActionRuleProcessorIT {
+
+  private static final int DELAY_ACTION_BY_SECONDS = 5;
```
Comment (liamtoozer) [inline src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java]: Unused?

File: src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+  @Value("${queueconfig.outbound-printer-queue}")
+  private String outboundPrinterQueue;
+
+  @Autowired private RabbitQueueHelper rabbitQueueHelper;
+  @Autowired private CaseRepository caseRepository;
+  @Autowired private UacQidLinkRepository uacQidLinkRepository;
```
Comment (liamtoozer) [inline src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java]: Unused?

File: src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  @Autowired private ActionRuleRepository actionRuleRepository;
+  @Autowired private ActionPlanRepository actionPlanRepository;
+  @Autowired private ActionRuleProcessor actionRuleProcessor;
```
Comment (liamtoozer) [inline src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java]: Unused?

File: src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+  @Value("${caseapi.port}")
+  private int caseApiPort;
+
+  @Value("${caseapi.host}")
+  private String caseApiHost;
+
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  @Rule public WireMockRule mockCaseApi = new WireMockRule(wireMockConfig().port(8089));
```
Comment (liamtoozer) [inline src/test/java/uk/gov/ons/census/action/schedule/ActionRuleProcessorIT.java]: Minor comment but is it worth adding a setUp method to purge the queues etc here? [Similar to this](https://github.com/ONSdigital/census-rm-case-processor/blob/8959ea6e147d7235b0bed94538d3a6a076248439/src/test/java/uk/gov/ons/census/casesvc/messaging/InvalidAddressReceiverIT.java#L57-L64)

**Pull request merged**